### ISSUE TITLE
running commands as www-user

### ISSUE THREAD

**Issue by levelos** [kubitron/redmine_git_hosting#5]
Trying to follow the instructions to run fetch_changsets() commands as the www user, in my case www-data. I'm using Apache / Passenger. Problem is when I run them as that user, I get all sorts of path not found issues, namely for rake and ruby. Any ideas? Thanks!

**Comment by kubitron** [comment]
Some questions: (1) are you running rake from the top-level of redmine?  (You cannot run it from the top of the plugin directory, since it won't find things), (2) do you have rake and ruby installed system wide, or are you using RVM?  Note that for "historical" reasons, you fetch changesets by running:

```
rake gitolite:fetch_changes
```

Note also that you need to make sure that RAILS_ENV is set properly.

Another way to run fetch_changesets is from the browser:

http://yourserver.com/sys/fetch_changes?key=xxxxxxxx

Here, the key is your "API key" on the settings/repositories page (if you have the WS option enabled on the same page).

**Comment by levelos** [comment]
Thanks John - I am running RVM. I actually went the browser route and that worked fine, although it didn't fix my issue unfortunately. All my change sets were correctly imported into the db though. 

I'm still a) getting prompted for a gitolite password as it seems the gitolite user's authorized_keys file isn't getting updated and b) if I add the correct entry in that file, than I get the access denied message sent from gitolite. No SSH auth errors, and the gitolite-admin repo is spot on and gets updated every time I update a user or project. Thanks again for your patience and help. This is what happens when a developer tries to pay sysadmin :)

**Comment by kubitron** [comment]
This sounds like the administrative key connection is broken.  What does the log file say?  (are you using my version of the plugin -- the error messages are much better).

So: do a fetch changesets operation, then post the resulting log message here.  (look into redmine's log/production.log) at the bottom, grab the last couple of entries (you can look at the beginning of each "paragraph" in the log).

**Comment by levelos** [comment]
Using your latest version John. Log entry follows:

> Processing SysController#fetch_changesets (for [IP ADDRESS] at 2012-01-25 13:36:33) [GET]
>   Parameters: {"key"=>"[KEY]"}
> Executing RESYNC_ALL operation on gitolite configuration
> Fetching changes from gitolite-admin repository to /tmp/redmine_git_hosting/gitolite/gitolite-admin
> Creating MD5 digests for Redmine Git Hosting hook
> Digest for post-receive.redmine_gitolite.rb: 08e9b3cf99fa6285331ccdcef8b9a0d3
> Our hook is already installed
> Completed in 44366ms (View: 1, DB: 1282) | 200 OK [https://[chiliserver.com]/sys/fetch_changesets?key=[KEY]]

**Comment by kubitron** [comment]
On 1/25/2012 1:41 PM, Lev Tsypin wrote:

> Using your latest version John. Log entry follows:
> 
> > Processing SysController#fetch_changesets (for [IP ADDRESS] at 2012-01-25 13:36:33) [GET]
> >    Parameters: {"key"=>"[KEY]"}
> > Executing RESYNC_ALL operation on gitolite configuration
> > Fetching changes from gitolite-admin repository to /tmp/redmine_git_hosting/gitolite/gitolite-admin
> > Creating MD5 digests for Redmine Git Hosting hook
> > Digest for post-receive.redmine_gitolite.rb: 08e9b3cf99fa6285331ccdcef8b9a0d3
> > Our hook is already installed
> > Completed in 44366ms (View: 1, DB: 1282) | 200 OK [https://[chiliserver.com]/sys/fetch_changesets?key=[KEY]]
> > 
> > ---
> > 
> > Reply to this email directly or view it on GitHub:
> > https://github.com/kubitron/redmine_git_hosting/issues/5#issuecomment-3659663
> > Ah.  This looks fine.

Are you sure that your intended user is authorized to access the 
repository?  Go to the settings page of the given project, then look at 
"members".  Members (or groups) need to be a least Developer or Manager 
to allow them to write and read from a repository.  They need to be at 
least a Reporter to read from the repository.

Note that gitolite manages the authorized_keys file automatically.  You 
should never have to edit it.  Note that keys do not go into the 
authorized_keys file unless there is at least one project that has that 
user in their members list!  Is it possible that you don't have your 
intended user as a member of the given project?

I find it useful to su to the gitolite user and cd to their home 
directory.  From there, you should be able to do a

"git clone git@yourserver.com:gitolite_admin.git"

 From there, you can look at the gitolite.conf file and look into the 
keys subdirectory to see if your keys are properly there... All of the 
redmine keys should be called "redmine_###_###.pub".  They will apear 
like that (without the pub) in the gitolite.conf file...

--KUBI--

## 

Professor John Kubiatowicz
673 Soda Hall
Berkeley, CA 94720

**Comment by kubitron** [comment]
OOps --- I meant:

git clone git@yourserver.com:gitolite-admin.git

**Comment by levelos** [comment]
Huh, when I try that, I get a prompt for the gitolite password? Message:

> Cloning into gitolite-admin...
> gitolite@chiliserver.com's password:

**Comment by kubitron** [comment]
Is gitolite the git user for you?  Usually it is just "git", unless you have changed something... If you don't try the above operation as the git user, then ssh won't know how to find the proper keys.  At any rate, that is a bit of a red  herring for you, since you are having problems.  Alternatively, just look into /tmp/redmine_git_hosting/.... for the gitolite.conf file and the keydir directory.  

What are the answers to my previous questions about membership?  Does it make sense that a user that is not in the membership of any project will not have their key in the authorized_keys file?  Can you tell me exactly what you were trying that wasn't working (that started this whole thread)?

**Comment by levelos** [comment]
The package install method installs gitolite for a user called gitolite rather than git and the Chili users definitely have the right role. To recap, everything working well except SSH access to the git repos. The gitolite-admin repo is getting updated with Chili users and their keys, I can use Smart HTTP, and changesets appear in the repo tab. Just can't get SSH access working. Thanks again for the help.

**Comment by kubitron** [comment]
Ignore smart HTTP. The authentication process has nothing to do with gitolite.

You say a couple of incompatible things.  If the gitolite-admin repo is getting updated with users and their keys, but you had to manually edit the authorized_keys file, this doesn't make sense.  That is not the way that gitolite behaves (nothing to do with the plugin).

Show me the gitolite.conf file, show me a list of the keydir directory, and show me the key that you are trying to use.  What does the authorized users file look like?  What is the exact error you get when you try to clone the repository?  Note, since you are not sure how to independently pull the gitolite-admin directory, make sure to delete the /tmp file and execute a fetch_changesets before showing me the above information...

**Comment by levelos** [comment]
Right. When I run gl-setup, what seems to be a correct entry gets added to the /var/lib/gitolite/.ssh/authorized_keys file. Then, if I run a fetch_changesets() operation or edit a Chili users account, that file isn't getting updated. Probably not a root issue. I have sudo/admin user on the server, in our case called developer. developer has generated a ssh key pair, named gitolite_admin_id_rsa and gitolite_admin_id_rsa.pub and that pair was used to run gl-setup, as the user gitolite. At this point, authorized_keys looks like:

# gitolite start

command="/usr/share/gitolite/gl-auth-command gitolite_admin_id_rsa",no-port-forwarding,no-X11-forwarding,no-agent-forwarding,no-pty ssh-rsa AAAAB3NzaC1yc2EAAAADAQA … developer@chiliserver.com

# gitolite end

I think so far so good. After I run a fetch_changesets() operation, this line is added:
command="/usr/share/gitolite/gl-auth-command id_rsa",no-port-forwarding,no-X11-forwarding,no-agent-forwarding,no-pty ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABAQCec/Et0z … developer@chiliserver.com

Not sure where that second key is coming from, maybe a concern. The production log when I run that looks clean:

> Processing SysController#fetch_changesets (for 199.223.125.115 at 2012-01-26 11:54:47) [GET]
>   Parameters: {"key"=>"[key]"}
> Executing RESYNC_ALL operation on gitolite configuration
> Fetching changes from gitolite-admin repository to /tmp/redmine_git_hosting/gitolite/gitolite-admin
> Our hook is already installed
> Completed in 25000ms (View: 1, DB: 958) | 200 OK [https://chiliserver.com/sys/fetch_changesets?key=[key]]

When logged into the server as developer, my gitolite admin, I get the same error as when trying to push/pull from the client:

> Cloning into gitolite...
> gitolite@code.thinkshout.com's password:   

My conf/gitolite.conf file looks good, here's a snippet:

> repo    gitolite-admin
>         RW+     =       gitolite_admin_id_rsa
> 
> repo    testing
>         RW+     =       @all
> 
> repo    wka/wkb
>         RW+     =       redmine_lev_1327376665_784119 redmine_sean_1327426981_314251 redmine_brandon_1327347107_560453 redmine_tauno_1327339882_614646  

and the keydir looks like:

> gitolite_admin_id_rsa.pub
> id_rsa.pub
> redmine_brandon_1327347107_560453.pub
> redmine_lev_1327376665_784119.pub
> redmine_sean_1327426981_314251.pub
> redmine_tauno_1327339882_614646.pub  

So the issues seems to be with gitolite / SSH, but I'm really stumped and not sure what's up. No issues in the ssh auth log.

**Comment by kubitron** [comment]
Ok.  One thing that seems strange here is the fact that the gitolite-admin key is gitolite_admin_id_rsa.  On my system, it is id_rsa.  In fact, I am the reason that id_rsa appears, since I attempt to restore the key to id_rsa.  

How do you run gl-setup?  Is it gl-setup ~/.ssh/id_rsa.pub?  My guess is that one of the problems that you have with pullling the gitolite-admin repo is that you don't have a ~/.ssh/id_rsa and ~/.ssh/id_rsa.pub key in the gitolite user homedirectory.

Also, where are you getting your conf/gitolite.conf file from?  Is it from /tmp/redmine_git_hosting/...?  It is possible that the push operation never succeeds.  As such, I am surprised that there are no errors in your log file.  When you run fetch_changesets, does it ever say:

"Committing changes to gitolite-admin repository"?

Are there any errors around it?  If you never see "Committing changes to the gitolite-admin repository", then there is something seriously wrong here.  Perhaps it is a bug in the plugin.

I am thinking that the plugin tries to add all of those key files and the configuration 
I know of several people that have had problems making Chili work with this plugin recently.  Perhaps there is an incompatibility that has crept into the process...

**Comment by kubitron** [comment]
Another thing here -- the fact that my plugin restablished the key tells me that there must be errors in the log.  

Do you not have _any_ errors in the log?  There must be errors...

**Comment by kubitron** [comment]
Another question: is there a .git file in ~gitolite/repositories/wka/wkb.git?  Given what you've told me, there should be a repository there already.

Also, try the following (by hand in the gitolite-admin directory that you pulled): edit the gitolite.conf file to add another repo with a key or two, commit the file, then push it back.  See whether there are any errors and whether the new repository gets created.

**Comment by levelos** [comment]
> Ok. One thing that seems strange here is the fact that the gitolite-admin key is gitolite_admin_id_rsa. On my system, it is id_rsa. In fact, I am the reason that id_rsa appears, since I attempt to restore the key to id_rsa.
> 
> How do you run gl-setup? Is it gl-setup ~/.ssh/id_rsa.pub? My guess is that one of the problems that you have with pullling the gitolite-admin repo is that you don't have a ~/.ssh/id_rsa and ~/.ssh/id_rsa.pub key in the gitolite user homedirectory.
> gitolite_admin_id_rsa is actually the key of a different user, in our case developer. I copied that key from /home/developer/.ssh to a /tmp directory, then ran gl-setup (as the gitolite user) on that. My understanding from the README was that the key of the gitolite-admin user was supposed to be named that. That's the same key that I copied to /var/www/chiliproject/.ssh. I installed gitolite using Debian's package method and at no point was an SSH key created for actual gitolite user, nor did I ever create one. Should I have?
> Also, where are you getting your conf/gitolite.conf file from? Is it from /tmp/redmine_git_hosting/...? It is possible that the push operation never succeeds. As such, I am surprised that there are no errors in your log file. When you run fetch_changesets, does it ever say:
> 
> "Committing changes to gitolite-admin repository"?
> 
> Are there any errors around it? If you never see "Committing changes to the gitolite-admin repository", then there is something seriously wrong here. Perhaps it is a bug in the plugin.
> I do see that, and changes seem to be pushed correctly. I'm able to access the gitolite-admin repo by checking it as the gitolite (or root) user using a file patch, E.g., git checkout /var/lib/gitolite/repositories/gitolite-admin.

**Comment by levelos** [comment]
Not in /chili-root/log/produciton.log. 

> Do you not have _any_ errors in the log? There must be errors...

**Comment by levelos** [comment]
There is, all our repos are already there as we're migrating them from Redmine/Gitosis. I've also tried creating a new project in Chili and repo does get created. 

## 

Lev Tsypin

**Comment by kubitron** [comment]
That is _very_ weird.  Is it possible that ~gitolite/.ssh is not writable by the gitolite user?  

Have you tried creating a project using only gitolite?  (that is what I was suggesting above).  Do the following to make things easier.  Copy your keys to ~gitolite/.ssh, i.e.

gitolite_admin_id_rsa => ~gitolite/.ssh/id_rsa
gitolite_admin_id_rsa.pub => ~gitolite/.ssh/id_rsa.pub

Make sure to chmod the new files as follows:

chmod 400 ~gitolite/.ssh/id_rsa
chmod 644 ~gitolite/.ssh/id_rsa
chmod 700 ~gitolite/.ssh

Then, su to gitolite.

Now, clone the admin directory with

```
git clone gitolite@yourserver.com:gitolite-admin.git
```

Cd into the conf file.  Edit the conf file by hand to add a new repo entry with a different name.  For instance:

```
repo mytest
        RW+ = redmine_lev_1327376665_784119
```

Then, commit this by hand to gitolite:

```
git commit -am "Testing"
git push
```

Now, look carefully at the resulting messages during the push.  Are there errors?  Does the new repository get created?  (There should be  ~gitolite/repositories/mytest.git).  Anything new in the authorized_keys file?  Every key in the .conf file should be put into the authorized_keys file.

**Comment by levelos** [comment]
Ok, perhaps some clues. First, though, I don't think the gitolite user itself needs to have any SSH key pairs. What I was able to do: 
- Login as my gitolite admin user, in our case developer.
- That user has an SSH key pair, ~/.ssh/gitolite_admin_id_rsa. I just used this name for the keys as that's what I understood the README to suggest. Note that these keys where the ones used when I ran gl-setup, and are the ones copied into the chili root .ssh directory.
- I need to run ssh-add ~/.ssh/gitolite_admin_id_rsa, likely because of the non-standard name.
- I can then clone the admin repo, git clone gitolite@chiliserver.com:gitolite-admin.

I made a change in the conf file, committed, and tried to push. This gave me the following error:

`Counting objects: 7, done.
Compressing objects: 100% (3/3), done.
Writing objects: 100% (4/4), 380 bytes, done.
Total 4 (delta 1), reused 0 (delta 0)
error: git-shell died of signal 13
fatal: The remote end hung up unexpectedly
error: error in sideband demultiplexer
To gitolite@chiliserver.com:gitolite-admin
   dd7f4d6..f6094ac  master -> master
error: failed to push some refs to 'gitolite@chiliserver.com:gitolite-admin'`

Some very quick Googling leads to a post-hook error of some sort, probably why the authorized_key file isn't being updated. The changes to the conf file ARE pushed, but looks like the commit hooks are not firing.

**Comment by kubitron** [comment]
On 1/27/2012 11:55 AM, Lev Tsypin wrote:

> Ok, perhaps some clues. First, though, I don't think the gitolite user itself needs to have any SSH key pairs. What I was able to do:
> - Login as my gitolite admin user, in our case developer.
> - That user has an SSH key pair, ~/.ssh/gitolite_admin_id_rsa. I just used this name for the keys as that's what I understood the README to suggest. Note that these keys where the ones used when I ran gl-setup, and are the ones copied into the chili root .ssh directory.
> - I need to run ssh-add ~/.ssh/gitolite_admin_id_rsa, likely because of the non-standard name.
> - I can then clone the admin repo, git clone gitolite@chiliserver.com:gitolite-admin.
> 
> I made a change in the conf file, committed, and tried to push. This gave me the following error:
> 
> `Counting objects: 7, done.
> Compressing objects: 100% (3/3), done.
> Writing objects: 100% (4/4), 380 bytes, done.
> Total 4 (delta 1), reused 0 (delta 0)
> error: git-shell died of signal 13
> fatal: The remote end hung up unexpectedly
> error: error in sideband demultiplexer
> To gitolite@chiliserver.com:gitolite-admin
>     dd7f4d6..f6094ac  master ->  master
> error: failed to push some refs to 'gitolite@chiliserver.com:gitolite-admin'`
> 
> Some very quick Googling leads to a post-hook error of some sort, probably why the authorized_key file isn't being updated. The changes to the conf file ARE pushed, but looks like the commit hooks are not firing.
> 
> ---
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/kubitron/redmine_git_hosting/issues/5#issuecomment-3693038
> Ok.  That's good -- it means that this is a problem with your gitolite 
> configuration, not the plugin.

Now -- are you _sure_ that your authorized_keys file is owned by the 
gitolite user and that it has a 600 mode?  Is .ssh owned by the gitolite 
user with 700 mode?  (I am wondering if you might have accidentally made 
this file owned by root).

--KUBI--

## 

Professor John Kubiatowicz
673 Soda Hall
Berkeley, CA 94720

**Comment by kubitron** [comment]
Any progress on this problem?  I'd love to be able to close out the issue...

**Comment by levelos** [comment]
Hey  John - Thanks for following up. Double checked and yes on both accounts. One underlying question I never quite understood and isn't clear in the docs to me, is what public key I should use to setup Gitolite? I install gitolite, and then I need to run gl-setup path/to/some/key.pub. But who's key should that be? In normal gitolite use cases, it would by my key, to be used on a client. But in the Chili case, who's key should be used? Thanks again!

**Comment by kubitron** [comment]
Hello.  The key that you use to setup Gitolite needs to be the partner of the key that you put in 

```
RAILS_ROOT/.ssh/gitolite_admin_id_rsa
```

and which is also in:

```
RAILS_ROOT/.ssh/gitolite_admin_id_rsa.pub
```

Think of this as an administrative key owned by Redmine to manage gitolite.  Now, when you execute "gl-setup", it doesn't have to be called that.  In fact, I just copy both the public and private keys into ~gitolite/.ssh/id_rsa and .ssh/id_rsa.pub.  and run gl-setup with that.  Then, when manually altering or checking on the configuration, the gitolite user can simply checkout the gitolite-admin repository using its own key.  For this key-pair, simply generate a new key using ssh-keygen.  Noone will use it except Redmine and (perhaps) the gitolite user when you su to them.

This is not the client key.  All client keys are copied into the user's my_account page.   Thus, you should put your personal public key into your account in redmine.  

Are you saying that you managed to resolve the issue so that I can close it?

**Comment by kubitron** [comment]
Also -- what was the ultimate issue?  Problem with the gitolite installation?

**Comment by levelos** [comment]
Unfortunately, still unresolved. I tried a variation of the setup as you mention, that is running gl-setup using the gitolite user. Then, I can `su gitolite` and checkout the gitolite admin repo. Much as when doing this as the developer user, when I push, the code gets pushed but I get the same error:

``` shell
error: git-shell died of signal 13
fatal: The remote end hung up unexpectedly
error: error in sideband demultiplexer
```

Could my hooks be messed up from all the variations I tried? What follows are the hooks located in ~/repositories/gitolite-admin.git.

post-update

``` shell
#!/bin/sh

# ensure that the admin is not sneaking in src/ and hooks/ :)
GIT_WORK_TREE=$GL_ADMINDIR git ls-tree --name-only master |
    perl -lne 'exit 1 if /^(src|hooks)$/' || {
        echo "*** ERROR ***" >&2
        echo "no files/dirs called 'src' or 'hooks' are allowed, sorry" >&2
        echo "until those files are deleted, the post-update hook will not run" >&2
        exit 1
    }

# checkout the master branch to $GL_ADMINDIR
# (the GL_ADMINDIR env var would have been set by gl-auth-command)
GIT_WORK_TREE=$GL_ADMINDIR git checkout -f master

od=$PWD
cd $GL_ADMINDIR
$GL_BINDIR/gl-compile-conf

cd $od

ADMIN_POST_UPDATE_CHAINS_TO=` cd $HOME;perl -e 'do ".gitolite.rc"; print $ADMIN_POST_UPDATE_CHAINS_TO'`
[ -n "$ADMIN_POST_UPDATE_CHAINS_TO" ] || ADMIN_POST_UPDATE_CHAINS_TO=hooks/post-update.secondary

if [ -f $ADMIN_POST_UPDATE_CHAINS_TO ] || [ -L $ADMIN_POST_UPDATE_CHAINS_TO ]
then
    exec $ADMIN_POST_UPDATE_CHAINS_TO "$@"
fi
```

post-receive

``` ruby
#!/usr/bin/env ruby

require 'digest/sha1'
require 'net/http'
require 'net/https'
require 'uri'

STDOUT.sync=true
$debug=false


def log(msg, debug_only=false, with_newline=true)
    if $debug || (!debug_only)
        print msg + (with_newline ? "\n" : "")
    end
end
def get_git_repository_config(varname)
    (%x[git config #{varname} ]).chomp.strip
end
def get_http_params(rgh_vars)
    clear_time   = Time.new.utc.to_i.to_s
    params = { "clear_time" => clear_time, "encoded_time" => Digest::SHA1.hexdigest(clear_time.to_s + rgh_vars["key"]) }
    rgh_vars.each_key do |v|
        if v != "key"
            params[v] = rgh_vars[v]
        end
    end
    params
end

def run_query(url_str, params, with_https)
    url_str = (with_https ?  "https://" : "http://" ) + url_str.gsub(/^http[s]*:\/\//, "")
    success = false
    begin
        url  = URI.parse(url_str)
        http = Net::HTTP.new( url.host, url.port )
        http.open_timeout = 20
        http.read_timeout = 180
        if with_https
            http.use_ssl = true
            http.verify_mode = OpenSSL::SSL::VERIFY_NONE
        end
        req  = Net::HTTP::Post.new(url.request_uri)
        req.set_form_data(params)
        response = http.request(req) do |response|
            response.read_body do |body_frag|
                success = true
                log(body_frag, false, false)
            end
        end
        #response = http.request(req)
        #puts response.header
    rescue Exception =>e
        #log("HTTP_ERROR:" + e.to_s, true, true)
        success = false
    end
    success
end




rgh_vars = {}
rgh_var_names = [ "hooks.redmine_gitolite.key", "hooks.redmine_gitolite.url", "hooks.redmine_gitolite.projectid", "hooks.redmine_gitolite.debug", "hooks.redmine_gitolite.asynch"]
rgh_var_names.each do |var_name|
    var_val = get_git_repository_config(var_name)
    if var_val.to_s == ""
        log("\n\nRepository does not have \"#{var_name}\" set. Skipping hook.\n\n", false, true)
        exit
    else
        var_name = var_name.gsub(/^.*\./, "")
        rgh_vars[ var_name ] = var_val
    end
end

$debug = rgh_vars["debug"] == "true"


# Let's read the refs passed to us
refs = []
$<.each  do |line|
    r = line.chomp.strip.split
    refs.push( [ r[0].to_s, r[1].to_s, r[2].to_s ].join(",") )
end
rgh_vars["refs[]"] = refs

if rgh_vars["asynch"] == "true"
    pid = fork
    exit unless pid.nil?
    pid = fork
    exit unless pid.nil?

    File.umask 0000

    STDIN.reopen '/dev/null'
    STDOUT.reopen '/dev/null', 'a'
    STDERR.reopen STDOUT

end

log("\n\n", false, true)
log("Notifying ChiliProject/Redmine project #{rgh_vars['projectid']} about changes to this repo...", true, true)
success = run_query(rgh_vars["url"], get_http_params(rgh_vars), true)
if !success
    success = run_query(rgh_vars["url"], get_http_params(rgh_vars), false)
end
if(!success)
    log("Error contacting ChiliProject/Redmine about changes to this repo.", false, true)
else
    log("Success", true, true)
    log("", true, true)
end
log("\n\n", false, true)

exit
```

**Comment by kubitron** [comment]
Hm... The post-receive hook is from the redmine_git_hosting plugin.  So, it is probably fine.  Not sure about the other one, but it is likely from the gitolite installation (not in a position to look at the moment).

Are there any errors in the system logs?  I'd look in /var/log/messages to see if there is an obvious problem.  Better yet, try starting:

```
tail -f /var/log/messages
```

Then, try pushing the repository.  That way you can see if there are error messages associate with the push.  Also, try looking at (say) "/var/log/secure" (if you have one).  You aren't using selinux, by any chance, are you?

Note that you have never answered my question about the authorized_keys file.  Is it owned by the gitolite user and writeable by them?

Perhaps it would be better to simply reinstall gitolite from scratch and get it working properly (so that you can push repos to it and the keys get inserted properly, etc.) before you try to connect it with Redmine/redmine_git_hosting?

**Comment by levelos** [comment]
> Are there any errors in the system logs? I'd look in /var/log/messages to see if there is an obvious problem. Better yet, try starting:
> 
> tail -f /var/log/messages
> 
> Then, try pushing the repository. That way you can see if there are error messages associate with the push. Also, try looking at (say) "/var/log/secure" (if you have one). You aren't using selinux, by any chance, are you?
> I did try and no useful messages in either log. It actually stated user gitolite logged in successfully, then closed the session. But clearly the login worked since the changes were pushed. Not using selinux, just ubuntu.
> 
> Note that you have never answered my question about the authorized_keys file. Is it owned by the gitolite user and writeable by them?
> Sorry, yes, they both are. 
> Perhaps it would be better to simply reinstall gitolite from scratch and get it working properly (so that you can push repos to it and the keys get inserted properly, etc.) before you try to connect it with Redmine/redmine_git_hosting?

I'll give it another go. Thanks again for all your help.

**Comment by kubitron** [comment]
Sorry I'm not more helpful.  Seems like you have some odd configuration problem....

**Comment by levelos** [comment]
You were very helpful and patient, thanks again. I'll update when we figure what the issue was.

**Comment by levelos** [comment]
So I found the solution, and I'm surprised other folks haven't run into this. The [install instructions I followed for Chili](https://www.chiliproject.org/projects/chiliproject/wiki/Installation_on_Ubuntu_10_10_v2) on Ubuntu run Chiliproject as the user chiliproject. The redmine_git_hosting instructions assume you're Chili user is www-data. In short, my sudoers file needed to contain chiliproject instead of www-data, and related file permissions.
